pkg/provider/cloud/openstack/provider.go

```diff
@@ -397,9 +389,16 @@ func (os *Provider) reconcileSecurityGroups(ctx context.Context, netClient *goph
 		if err != nil {
 			return nil, fmt.Errorf("failed to create security group: %w", err)
 		}
+		cluster, err = update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
+			kubernetes.AddFinalizer(cluster, SecurityGroupCleanupFinalizer)
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to add security group finalizer: %w", err)
+		}
```

Comment:
Finalizers should be added before the security groups are created. Doing it this way will potentially leak SGs when cleaning up.

I also don't understand how this fixes the linked issue. Even with this change, once all reconciling is done, the cluster would still have the finalizer and during cleanup we still do not know whether we created the SG or not.

What the AWS provider does is annotate AWS resources it creates, precisely because we need to know whether something was created by KKP or not. Does OpenStack support something similar, so KKP can claim ownership?

Reply:
I can see how we could also use the existence of the finalizer as a signal whether we own the SG, but

* it should still be added before OpenStack is modified
* we support multiple SGs and so using a single finalizer won't work as an ownership signal